app/src/main/java/com/dmonzonis/nookpendium/CollectionActivity.kt

```diff
@@ -79,64 +83,87 @@ class CollectionActivity : AppCompatActivity(), SortDialogFragment.SortDialogLis
                     fabFilters.show()  // Reappear on scroll up
                 }
             }
         })
+    }
 
-        fabFilters.setOnClickListener { setFilterButtonsEnabled(!isFilterSubmenuOpen) }
-        setFilterButtonsEnabled(false)
-        fabFilterThisMonth.setOnClickListener { filterByThisMonth() }
-        fabFilterClear.setOnClickListener {
-            recordset.restore()
-            viewAdapter.setRecords(recordset.records)
-        }
-        fabSortBy.setOnClickListener {
-            val sortByDialog = SortDialogFragment()
-            sortByDialog.show(supportFragmentManager, "sort_by")
-        }
+    // Create toolbar menu
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menuInflater.inflate(R.menu.toolbar_menu, menu)
+        return super.onCreateOptionsMenu(menu)
     }
 
-    override fun onDialogPositiveClick(dialog: DialogFragment, token: String, descending: Boolean) {
-        viewAdapter.setRecords(recordset.applySort(token, descending))
+    // Restores the recordset to the original, computes all active filters
+    // and the current sorting on it, and then updates the view adapter to use the new
+    // recordset
+    fun recomputeFilters() {
+        recordset.restore()
+        // First apply all filters, and then sort the resulting recordset
+        filterManager.applyFilters(recordset)
+        sortManager.applySort(recordset)
+        viewAdapter.setRecords(recordset.records)
     }
 
-    private fun setFilterButtonsEnabled(enabled: Boolean) {
-        isFilterSubmenuOpen = if (enabled) {
-            fabFilters.setImageResource(R.drawable.ic_clear_black_24dp)
-            fabFilterThisMonth.show()
-            fabFilterClear.show()
-            fabSortBy.show()
-            cardviewFilterThisMonth.visibility = View.VISIBLE
-            cardviewFilterClear.visibility = View.VISIBLE
-            cardviewSortBy.visibility = View.VISIBLE
+    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
+        // Close all drawers before to avoid having multiple drawers opening on top of each other
+        closeAllDrawers()
+        return if (drawerToggle.onOptionsItemSelected(item)) {
             true
-        } else {
-            cardviewFilterThisMonth.visibility = View.GONE
-            cardviewFilterClear.visibility = View.GONE
-            cardviewSortBy.visibility = View.GONE
-            fabFilterThisMonth.hide()
-            fabFilterClear.hide()
-            fabSortBy.hide()
-            fabFilters.setImageResource(R.drawable.ic_search_black_24dp)
-            false
+        } else when (item?.itemId) {
+            R.id.miFilterButton -> {
+                if (drawerLayout.isDrawerOpen(navFilterMenu))
+                    drawerLayout.closeDrawer(navFilterMenu)
+                else
+                    drawerLayout.openDrawer(navFilterMenu)
+                true
+            }
+            else -> super.onOptionsItemSelected(item)
```

Comment:
this is super hard to read and it's because you're using multiple navigationView which are not designed for that and on top of that it's set as the option menu.

for now I guess we can do with it, but I think it's gonna be error-prone and a lot of maintenance down the road, it already is:

1. if you try swiping from left to right while the filter menu is open (to open the navigation menu), you have to do it slow enough for the other menu to close, otherwise both "conflict" and both close

2. if you open the filter menu, the nav menu's icon turns into an arrow, this is already weird but ok let's assume that's normal, when you click on it it doesn't *just* close the filter menu, it opens the navigation menu, and then the icon with the navigation menu open becomes the burger icon when it should be the back arrow icon, and then if you go back to the filter menu by swiping, the burger menu icon stays, making the UI completely inconsistent, almost random

Reply:
yes, I know it's messy, but to fix it we'll have to remove the current ActionBar by a custom Toolbar widget, which I'll do in another PR hopefully this weekend